### DISCUSSION TITLE
Blade Displacement Response and Aerodynamic Damping

### DISCUSSION THREAD

**Discussion by peterfakhry** [OpenFAST/openfast#1159] [Q&A] [answered]
Hello there, so i have been stuck for months trying to figure out why there is a bias between my produced numerical model and `FAST` regarding blade response (i get perfect agreement between the curves when i manually add a `0.35m` in the in plan and `0.5m` in the out of plan responses for all the three blades). As you can see that the `FAST `response never goes to the negative side (does that mean wind loading is much higher than gravity effect?). Also i have noticed an important thing, the response frequency content is sensitive to aerodynamic damping in my model. How is aerodynamic damping is implemented in `FAST`?.
![image](https://user-images.githubusercontent.com/40067011/175771071-c658f827-bc38-4610-839b-4b9cd884a070.png)

**Comment by jjonkman** [comment]
Dear @peterfakhry,

Presumably the "Numerical" result is from your own custom solver; can you clarify?

From the plots you show, I would guess that the aerodynamic loads are higher in FAST than in your solver.  I would suggest =first comparing the models in the absence of blade flexibility.  How do the models compare for rigid blades?

Aerodynamic damping is not written as a distinct term in FAST / OpenFAST.  Instead, the aerodynamic loads are computed based on the relative velocity between the wind and the blade aerodynamic nodes, including elastic velocity.  So, aerodynamic damping is intrinsic to the aero-elastic coupling.

Best regards,

**Reply by peterfakhry** [reply]
Thank you Professor for your reply. 
- Yes, this is my custom solver, i have developed it using MATLAB following the Lagarangian model in "Vibration Control of Wind Turbines with Variable Rotor Speed and Grid Faults" (Staino et al. 2013). 
- I have adjusted the `AdjFlSt` and `AdjEdSt` (in `NRELOffshrBsline5MW_Blade`) to `11`. This is the maximum value I could use before the FAST runs into a Fatal Error. Both models are subjected to a steady `12m/sec` wind speed. I have multiplied my blade stiffness by 11 as well and got the below curves.  
![untitled](https://user-images.githubusercontent.com/40067011/176427433-bcc07356-11e6-4ff4-a0d1-f32f63ff72cc.jpg)
![untitled_1](https://user-images.githubusercontent.com/40067011/176427486-13498a97-5027-417b-ae9f-68ad93c2148a.jpg)

**Reply by jjonkman** [reply]
Dear @peterfakhry,

In ElastoDyn, `AdjFlSt` and `AdjEdSt` are simple linear scaling factors on the specified flapwise and edgewise stiffness distributions.  If you change these much different from unity, I would expect that you'd have to recompute the mode shapes accordingly.  Likewise, increasing these scaling factors much more than unity will increase the blade natural frequencies, which may eventually require you to drop the ElastoDyn (and glue code) time step.

Again, I would suggest first comparing the models in the absence of blade flexibility. How do the aerodynamic loads compare for rigid blades?

Best regards,

**Reply by peterfakhry** [reply]
I guess that increasing these factors would result in higher stiffness hence rigid blades and absence of flexibility.

**Reply by jjonkman** [reply]
I can't speak for your solver, but in ElastoDyn, it is better to disable the blade degrees of freedom to make the rotor rigid rather than increasing the elastic stiffness, which would require smaller time step. You'll never make the rotor perfectly rigid by increasing the stiffness.  You could also run AeroDyn standalone to compute the aerodynamic loads for a rigid turbine.

Best regards,

**Reply by peterfakhry** [reply]
I turned off all the degrees of freedom regarding the FAST and multiplied my stiffness in my numerical model by 1000. As expected Professor, all results are zeros however i got again negative values in my model.
![untitled](https://user-images.githubusercontent.com/40067011/176448697-9e0d471a-9ac4-4887-aebc-0ccbfb503cdf.jpg)

**Reply by jjonkman** [reply]
Now that the deflections are zero (or near zero in your case), how do the applied aerodynamic loads compare?

**Reply by peterfakhry** [reply]
Thanks a lot for your efforts Professor and sorry for late reply. I get my aerodynamic loads from BEM equations in terms of normal (out of plan) and  tangential (in plan) using the following formulas (Hansen 2015):
![image](https://user-images.githubusercontent.com/40067011/176791871-d17ee60a-54e1-4119-82fc-bcdea91ba1ca.png)
Finally i multiply the `PN `by the flapmode shape (expressed in terms of `x` along the blade length) and the `PT `by the edgewise modeshape and integrate along from `0 `to `Lb `to get a modal load as (where `phi `as the respective modeshape):
![image](https://user-images.githubusercontent.com/40067011/176792211-7933a16e-e3fb-480e-b0b6-8a59075c1240.png)
As for the tower, i get the resultant of the three blades (tangential) in the edgewise direction and multiply by the respective cosine azimuth angle as
![image](https://user-images.githubusercontent.com/40067011/176792404-210c7dcc-ffd6-4d9a-8117-09ea76e29ff4.png)
For the fore-aft direction , i get the resultant of the three blades (normal) as
![image](https://user-images.githubusercontent.com/40067011/176792502-41f6e87a-830f-4dbd-8fff-f094e3e72e84.png)
Is that compare to "RootFxb1" and RootFyb1" in FAST and how to run AeroDyn to compute the aerodynamic loads.

**Reply by jjonkman** [reply]
Your equations sound reasonable for a case without precone, tilt, deflection, etc, but there are many details in any BEM formulation.

ElastoDyn outputs `RootFxb1` and `RootFyb1` are reaction loads that include not only the applied aerodynamic loads, but also the weight/inertia of the blade.  I would suggest comparing your pure aerodynamic loads, e.g. `p_N(x)` and `p_T(x)` in your nomenclature, against the equivalent loads computed and output by AeroDyn (e.g., `BαNβFx` and `BαNβFy` for output node `Nβ` of blade `Bα`).

Best regards,

**Reply by peterfakhry** [reply]
Thanks a lot Professor for your reply i really appreciate your help.
I have followed the steps in AeroDyn manual to run it in a standalone mode. Also have adjusted the `.dvr` and `.dat` files to match my case and got the following results 
![image](https://user-images.githubusercontent.com/40067011/177055997-b390c1be-4247-4410-a130-2ce954f42f7d.png)
I can see that they are not changing over time since the wind speed was set to `12m/sec` with shear exp `0 `and have adjusted the turbine dimensions to match my case. According to my knowledge the all the `Fx `corresponds to the out of plan forces and the `Fy `correspond to the in plan froces. Now i have assumed that the `9 `nodes are equally spaced along a `61.5m `blade (which is not the case as seen in Figure 3. in the AeroDyn manual) and have generated a polynomial to the 4th degree between the distance along the blade and the aerodynamic load. For each time step multiply the later value by the corresponding modeshape and integrate from `0 `to `Lb` as in 
![image](https://user-images.githubusercontent.com/40067011/177056262-0ffdef15-fd59-425c-9c6c-dee166696668.png)

 Below are the load values of my numerical model and the AeroDyn. There is some sort of matching in the in plan loads however out of plan loads are far away. 
- Why there are just `9 `nodes although there are `17 `specified in the ElastoDyn of FAST?
- What is the appropriate spacing between these nodes as this will help much in generating the polynomial expression?

![untitled](https://user-images.githubusercontent.com/40067011/177056205-e0a9ad97-dc55-4449-8453-d357575e8436.jpg)

**Comment by ebranlard** [comment, marked as answer]
Hi @peterfakhry 

Are you comparing the integrated loads at the blade root in the figure you enclosed? 

As Jason recommended, I agree that you might have to match the aerodynamic loads first, and it will help to plot Fx and Fy as function of the blade radius r for OpenFAST and your code. 

Regarding your questions about radial locations:
- AeroDyn and ElastoDyn use different number of nodes and node locations for the user convenience, and there is no limit on the number of nodes you can use. OpenFAST takes care of "mapping" the loads and motions between the mesh of AeroDyn and ElastoDyn.
- ElastoDyn will always have equidistant node locations, based on the parameter (`BldNodes`). You can find the formula for the nodes [here](https://github.com/ebranlard/welib/blob/master/welib/fast/postpro.py#L34). ElastoDyn uses a mid-point rule when performing integrals, so you'll find differences with your method if you use trapezoidal integration and different node locations. You might need at least 100 nodes before the differences between the two approaches (mid-point or trapezoidal) is negligible. Most users use few radial stations for ElastoDyn, but I would personally recommend 40 or more to ensure "convergence" of the integrals. This will increase the computational time though. 
- AeroDyn nodes do not need to be equidistant. Some people choose a cosine distribution to have more "resolution" towards the tip and root, where the load gradients are usually largest. I would nevertheless recommend to use an equidistant spacing with about 40-50 nodes.   For your debugging, you can use fewer stations of course. 
- The "old nodal outputs" of AeroDyn and ElastoDyn are limited to 9 radial locations which are selected amongst the set of radial locations of each module. We have now introduced "new nodal outputs" at the end of the input files to output quantities at all the radial locations of the module.  See [here](https://openfast.readthedocs.io/en/dev/source/user/aerodyn/input.html#nodal-outputs) for AeroDyn. You can use these outputs to compare the radial loads obtained between your BEM code and AeroDyn. 

Here's some additional recommendations for the aerodynamics:
- Simplify your blade: no prebend, presweep, precone (this will simplify the aerodynamics, and the structure. With prebend/cone etc, OpenFAST might use a different coordinate system than you do for your aerodynamic calculation)
- Compare without inductions first (WakeMod=0, a=0, a'=0). The angle of attack is then the geometrical angle of attack and it should match exactly. Then if the angle of attack match, the Cl, Cd, Cm should also match exactly. 
- Introduce inductions (WakeMod=1), and compare a and a', then Fx and Fy. 

Looking at the elasticity, some other sources of differences might arise, here are some additional considerations:
- As I mentioned above, the nodal locations, and the integration method will affect your integrals.  
- ElastoDyn uses "twisted" shape functions, where the structural twist is used to combine the flap and edge shape functions. You might need to remove the structural twist in the ElastoDyn file at first when comparing with your method. This can be a large source of differences. 
- Also remove precone, prebend, presweep from the blade file at first, as this might be handled differently in your code and ElastoDyn.
- ElastoDyn includes non-linear geometrical stiffening corrections that are likely not included in your model.
- You can try to match your structural model with ElastoDyn by running OpenFAST simulations without AeroDyn, using some initial values of the in-plane and out-of-plane deflections, with or without rotation. 

I hope that helps,

Emmanuel

**Reply by peterfakhry** [reply]
Thanks a lot for your help. I really appreciate your and Professor Jonkman efforts
